crates/bitwarden-ipc/src/ipc_client.rs

```diff
@@ -72,26 +140,75 @@ where
 
     /// Receive a message, skipping any messages that cannot be deserialized into the expected
     /// payload type.
-    pub async fn receive_typed<Payload>(
+    async fn receive_typed<Payload>(
         &self,
+        receiver: &Com::Receiver,
         timeout: Option<Duration>,
     ) -> Result<
         TypedIncomingMessage<Payload>,
         TypedReceiveError<
             <Payload as TryFrom<Vec<u8>>>::Error,
             Crypto::ReceiveError,
-            Com::ReceiveError,
+            <Com::Receiver as CommunicationBackendReceiver>::ReceiveError,
         >,
     >
     where
         Payload: TryFrom<Vec<u8>> + PayloadTypeName,
     {
         let topic = Some(Payload::name());
-        let received = self.receive(topic, timeout).await?;
+        let received = self.receive(receiver, &topic, timeout).await?;
         received.try_into().map_err(TypedReceiveError::Typing)
     }
 }
 
+impl<Crypto, Com, Ses> IpcClientSubscription<Crypto, Com, Ses>
+where
+    Crypto: CryptoProvider<Com, Ses>,
+    Com: CommunicationBackend,
+    Ses: SessionRepository<Session = Crypto::Session>,
+{
+    /// Receive a message, optionally filtering by topic.
+    /// Setting the timeout to `None` will wait indefinitely.
+    pub async fn receive(
+        &self,
+        timeout: Option<Duration>,
+    ) -> Result<
+        IncomingMessage,
+        ReceiveError<
+            Crypto::ReceiveError,
+            <Com::Receiver as CommunicationBackendReceiver>::ReceiveError,
+        >,
+    > {
+        self.client
+            .receive(&self.receiver, &self.topic, timeout)
+            .await
+    }
+}
+
+impl<Crypto, Com, Ses, Payload> IpcClientTypedSubscription<Crypto, Com, Ses, Payload>
+where
+    Crypto: CryptoProvider<Com, Ses>,
+    Com: CommunicationBackend,
+    Ses: SessionRepository<Session = Crypto::Session>,
+    Payload: TryFrom<Vec<u8>> + PayloadTypeName,
+{
+    /// Receive a message.
+    /// Setting the timeout to `None` will wait indefinitely.
+    pub async fn receive(
+        &self,
+        timeout: Option<Duration>,
+    ) -> Result<
+        TypedIncomingMessage<Payload>,
+        TypedReceiveError<
+            <Payload as TryFrom<Vec<u8>>>::Error,
+            Crypto::ReceiveError,
+            <Com::Receiver as CommunicationBackendReceiver>::ReceiveError,
+        >,
+    > {
+        self.client.receive_typed(&self.receiver, timeout).await
```

Comment:
I wonder if we should inline the `receive_typed` method from `IpcClient` here, it's only used here and it's simple enough:
```
    pub async fn receive(
        &self,
        timeout: Option<Duration>,
    ) -> Result<
        TypedIncomingMessage<Payload>,
        TypedReceiveError<
            <Payload as TryFrom<Vec<u8>>>::Error,
            Crypto::ReceiveError,
            <Com::Receiver as CommunicationBackendReceiver>::ReceiveError,
        >,
    > {
        let topic = Some(Payload::name());
        let received = self.client.receive(&self.receiver, &topic, timeout).await?;
        received.try_into().map_err(TypedReceiveError::Typing)
    }
```

Reply:
I was thinking that the req/res would build upon this and so it made sense to keep it separate, but I ended up inlining it in a follow up PR so maybe we can just merge this as is 🤷‍♂️